### ISSUE TITLE
Exception: Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host

### ISSUE THREAD

**Issue by sraddhan** [JonasSyrstad/Stardust.Paradox#15]
**Description**

While package does fantastic job almost all of the times, every once in a while it starts throwing error and doesn't recover from there until I stop and start the Azure Function App Service. I have provided entire stack trace in the `Additional Context` section. 

After spending several days trying to debug/reproduce the issue, I found that this was an actual bug in Gremlin.Net which according to the repository info should now be fixed after t[his PR](https://github.com/apache/tinkerpop/pull/1016) was merged. However, I still see the issue popping up. 

I have tried to catch the exception and Dispose the context there by forcing new connection. However, issue persists and can only be corrected by restarting the Azure App Service (Function App)

**To Reproduce**
This is sporadic. Not able to reproduce when desired. It does happen at infrequent intervals. Some times once a day or once in 5 days. 

**Expected behavior**
If there's an issue in connection it should recover. 

**Environment**
- Azure Functions 2.0. 
- C#
- .Net Standard 2.0 class library uses Stardust.Paradox to execute Gremlin queries. 

**Additional context**
----------------------------------------------------------
Please find following example code, actual exception message and stack trace.

**Example Code** 
`await query.Count().ExecuteAsync();`

**Exception Message** 
`Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host`

**Stacktrace** 

{System.IO.IOException: Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host. ---> System.Net.Sockets.SocketException: An existing connection was forcibly closed by the remote host --- End of inner exception stack trace --- at System.Net.Security.SslStreamInternal.<WriteSingleChunk>g__CompleteAsync|36_1[TWriteAdapter](ValueTask writeTask, Byte[] bufferToReturn) at System.Net.Security.SslStreamInternal.<WriteAsyncInternal>g__ExitWriteAsync|35_0[TWriteAdapter](ValueTask task) at Gremlin.Net.Driver.WebSocketConnection.SendMessageAsync(Byte[] message) at Gremlin.Net.Driver.Connection.SendAsync(RequestMessage message) at Gremlin.Net.Driver.Connection.SubmitAsync[T](RequestMessage requestMessage) at Gremlin.Net.Driver.ProxyConnection.SubmitAsync[T](RequestMessage requestMessage) at Gremlin.Net.Driver.GremlinClient.SubmitAsync[T](RequestMessage requestMessage) at Gremlin.Net.Driver.GremlinClientExtensions.SubmitAsync[T](IGremlinClient gremlinClient, String requestScript, Dictionary`2 bindings) at Stardust.Paradox.Data.Providers.Gremlin.GremlinNetLanguageConnector.ExecuteAsync(String compileQuery) at Stardust.Paradox.Data.Traversals.GremlinQuery.ExecuteAsync() at .Services.Common.StoryHelper.GetStoryMetaData(String status, Int32 page, Int32 size, GraphContext context, ILogger log) in \Services\Common\StoryHelper.cs:line 476 at .Services.Providers.StoryGraphProvider.GetStories(String status, Int32 page, Int32 size) in \Services\Providers\StoryGraphProvider.cs:line 71}
Data:

{System.Collections.ListDictionaryInternal}
HResult: -2146232800
HelpLink: null
IPForWatsonBuckets: {1646180022}
InnerException: {System.Net.Sockets.SocketException (10054): An existing connection was forcibly closed by the remote host}
IsTransient: false
Message: "Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host."
RemoteStackTrace: null
Source: "System.Private.CoreLib"
StackTrace: " at System.Net.Security.SslStreamInternal.<WriteSingleChunk>g_CompleteAsync|36_1[TWriteAdapter](ValueTask writeTask, Byte[] bufferToReturn)\r\n at System.Net.Security.SslStreamInternal.<WriteAsyncInternal>g_ExitWriteAsync|35_0[TWriteAdapter](ValueTask task)\r\n at Gremlin.Net.Driver.WebSocketConnection.SendMessageAsync(Byte[] message)\r\n at Gremlin.Net.Driver.Connection.SendAsync(RequestMessage message)\r\n at Gremlin.Net.Driver.Connection.SubmitAsync[T](RequestMessage requestMessage)\r\n at Gremlin.Net.Driver.ProxyConnection.SubmitAsync[T](RequestMessage requestMessage)\r\n at Gremlin.Net.Driver.GremlinClient.SubmitAsync[T](RequestMessage requestMessage)\r\n at Gremlin.Net.Driver.GremlinClientExtensions.SubmitAsync[T](IGremlinClient gremlinClient, String requestScript, Dictionary`2 bindings)\r\n at Stardust.Paradox.Data.Providers.Gremlin.GremlinNetLanguageConnector.ExecuteAsync(String compileQuery)\r\n at Stardust.Paradox.Data.Traversals.GremlinQuery.ExecuteAsync()\r\n 
at .Services.Common.StoryHelper.GetStoryMetaData(String status, Int32 page, Int32 size, GraphContext context, ILogger log) in \\Services\\Common
StoryHelper.cs:line 476\r\n at .Services.Providers.StoryGraphProvider.GetStories(String status, Int32 page, Int32 size) in \\Services\\Providers
StoryGraphProvider.cs:line 71"
TargetSite: {Void Throw()}
WatsonBuckets: null
_HResult: -2146232800
_className: "System.IO.IOException"
_data: {System.Collections.ListDictionaryInternal}
_dynamicMethods: null
_exceptionMethod:

{Void Throw()}
_helpURL: null
_innerException:

{System.Net.Sockets.SocketException (10054): An existing connection was forcibly closed by the remote host}
_ipForWatsonBuckets:

{1646180022}
_message: "Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host."
_remoteStackIndex: 0
_remoteStackTraceString: null
_source: "System.Private.CoreLib"
_stackTrace:

{sbyte[1536]}
_stackTraceString: null
_watsonBuckets: null
_xcode: -532462766
_xptrs:

{0}

**Comment by JonasSyrstad** [comment]
Thanx for the feedback. I'll look into it and see if i can find a way to recover from this

**Comment by JonasSyrstad** [comment]
I have done  some changes to try to mitigate the issue. But i suspect that the problem is related to cosmosDB and gremlin.net. I saw that they did some changes to the connection pool in 3.4.1

See for details:
https://issues.apache.org/jira/browse/TINKERPOP-2090

**Comment by sraddhan** [comment]
@JonasSyrstad - Thank you for taking a look at this. I updated the package at my end but haven't been able to test this.  Reason: After updating the package observed that query evaluation is broken. Don't want to create another issue as I know this might be a minor fix at your end. I have explained the issue below.

-----------------------------------------------------------------------------------------------
None of the Gremline queries are evaluated correctly. I have debugged and provided the reason below. 

`await query.Count().ExecuteAsync()` code results in error. 

Error Message: `ServerError: Gremlin Malformed Request: Unable to deserialize the request to a valid Gremlin Request.`

While debugging the code found that  `query` resolves in parameter index from the parameter collection. Rather than actual parameter value. 

`G.V().HasLabel(VertexType.product.ToString())` resolves to 

Actual:
`{g.V().hasLabel(__p0)}` 

Before: 
`{g.V().hasLabel("product")}`

 
I had a chance to look through the repository/source code. I see that, in GremlinQuery class, the query is formatted via parameterized dictionary. So, seems `__p0` is expected and was done so that query along with parameters can be passed to a method _client.SubmitAsync (in Gremlin.Net driver ). 

However, the fact remains; it does break and error message is consistent.  

Seems above change was done as part of V2 merge?

**Comment by JonasSyrstad** [comment]
I tested this and cannot reproduce your issue. Can you check the query before it is executed and see what the content of the Parameters property?

**Comment by sraddhan** [comment]
I saw parameter Dictionary populated. There were total 5 parameters (in my case) Further checked the values too, they seem to be correctly set as I had passed them while building the query.

**Comment by sraddhan** [comment]
Narrowed it down to where it fails: 

Works:  `{g.V().hasLabel(__p0).has(__p1,__p2).order().by(__p4,decr)}`

Fails :  `{g.V().hasLabel(__p0).has(__p1,__p2).has(__p3,gte(__p4)).order().by(__p5,decr)}`

I believe  `.has(__p3,gte(__p4))` part is making the query to fail.

**Comment by sraddhan** [comment]
@JonasSyrstad - if you could please confirm if queries with `.has(__p3,gte(__p4))` is working (or not working) at your end that would be very helpful!

**Comment by JonasSyrstad** [comment]
After some digging it turns out that CosmosDb doesn't know how to deal with numbers larger than int.MaxValue when deserializing numbers. 
I have made an exception in the ComposeParameter method that will skip parameterization of numbers larger than intMaxValue

**Comment by sraddhan** [comment]
Thanks! 

Updated the package and I can confirm that it has fixed the query issue. 
For the issue https://github.com/JonasSyrstad/Stardust.Paradox/issues/15, I will continue to observe for this week and update here.

**Comment by sraddhan** [comment]
Update: It worked beautifully the whole day. However, after ~150 requests, the connection died with following error:

`All 4 connections have reached their MaxInProcessPerConnection limit of 32. Consider increasing either the PoolSize or the MaxInProcessPerConnection limit.`

And almost every requests after that returned following error:

`No connection to the server available which most likely means that the server is completely unavailable.`

Had to restart the Azure Function App to get it back working.

**Comment by sraddhan** [comment]
Additional Info - Using JMeter, made 10 concurrent simultaneous requests (more than 250 total within less than 5 minutes), all reported success. 

Earlier when it failed, ~150 requests were made across the span of 10 hours.

**Comment by JonasSyrstad** [comment]
I added 2 static properties on GremlinNetLanguageConnector that allows you to controll the server connection:
ConnectionPoolSettings:  PoolSize and MaxInProcessPerConnection. 
WebSocketConfiguration: set KeepAliveInterval and the like.
I'm also trying to add better handling of connection errors. I have been able to reproduce the issue by making my machine hibernate. 

from this, https://issues.apache.org/jira/browse/TINKERPOP-2090, thread it seems like cosmosdb is closing connections  

From my tests of Stardust.Paradox.Data.Providers.Gremlin v 2.1.1 the connection will be reset when connection errors are raised. Hope this solves your stability issues.

**Comment by sraddhan** [comment]
Thanks Jonas. I updated the package and redeployed. Still see the error occur. Happened 3 times today.
In one instance, after Azure Function App restart, this error occurred after 8 successful requests only. 

`All 4 connections have reached their MaxInProcessPerConnection limit of 32. Consider increasing either the PoolSize or the MaxInProcessPerConnection limit.`

and

`No connection to the server available which most likely means that the server is completely unavailable.`

JMeter tests executed successfully (without any issues). Ran concurrent requests (10 threads) and total 1000 in less than 5 minutes and it worked without any issue. 

-----------------------------------------

**Update** - Please wait on checking above. I think I have found a possible issue in my code. I will work on that and will confirm if above still holds true.

**Comment by sraddhan** [comment]
**After code fix (Details given at the end)** 

- ~Now, I no longer have to restart the Azure Function App after it encounters the error.~ 
- After no requests for about 12 hours to the Azure Function App. All the requests to the App Server started failing. Had to restart! So, issue still persists. 
- However, errors still continue to occur. Had ~4% error rate for 30 K requests in 2 hours. 

![image](https://user-images.githubusercontent.com/7027023/55504372-b6f49280-561e-11e9-90cb-b1f57bcccc85.png)
----

**Error messages** 

----

`The remote party closed the WebSocket connection without completing the close handshake. Unable to read data from the transport connection: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond. A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond`

----

`The WebSocket is in an invalid state ('CloseSent') for this operation. Valid states are: 'Open, CloseReceived'` 

----

**Code fix details** 

- Found out that `Context` in my code continued to live. Reason - `Context` was configured as `AddScoped` in my dependency injection code. Changed it to `AddTransient` so that it's new per request.
- ~After above fix, at least, I don't have to restart my the Function App after error.~
- Above is not true. Had to restart the server the next day.
- I know `using` syntax is preferred but since `Context` is injected in constructor in my project, decided to rely on DI for now.

**Comment by sraddhan** [comment]
**Update** 

I think Azure Cosmos DB was much more unstable than normal this morning. There were lot of connections getting dropped. In the afternoon things got back to normal and connection was recovering itself. 

I continue to see ~4% error rate for 30 K requests.

**Comment by JonasSyrstad** [comment]
Good to hear that the connection recovers. I dont really understand why you get 4% error rate, but hopefully with version 2.1.2 it will not give you any interruption of service. 
The underlying issue seems to be related to germlin.net and cosmosdb and i dont think there is more I can do from this level of abstraction.
I'll close this issue, but feel free to reach out if there is anything.
Thank you for reporting and following up :)

**Comment by sraddhan** [comment]
Sounds good. Thanks for your prompt attention and turn around!
